nntrainer/optimizers/optimizer.cpp

```diff
@@ -58,6 +58,9 @@ void Optimizer::apply_gradients(std::vector<Weight> &weight_list,
     if (!weight.getTrainable())
       continue;
 
+    /** calculate regularization gradient before applying the gradient */
+    weight.calcRegularizationGradient();
```

Comment:
I wonder if this logic can go inside `weight::apply_gradient`.

Reply:
Adam takes in the gradient and updates the gradient which is to be applied to the weight.
So, the regularization gradient must be added before adam operation updates the weight.

Moving it inside the `weight::apply_gradient` would make regularization skip adam. This might be useful for `weight_decay` (weight decay skips adam operations on those gradient) when we support it later.